### ISSUE TITLE
How are negative anchors selected?

### ISSUE THREAD

**Issue by barisbatuhan** [Megvii-BaseDetection/YOLOX#725]
In your paper and also in this repository, I can see that you apply dynamic-k-matching to get the indices of positive matched anchors. However, how do you select negative anchors (in other words background anchors)? Do you automatically assign all remaining/unmatched anchors as negative samples? In many detector algorithms, some strategies such as OHEM are used and usually, number of positive and negative anchors are balanced with that way (like getting "num_positive_anchors x 3" times negative anchors). Which strategy is followed to select negative anchors and in which part of the YOLOX_Head class can I find that structure? If all remaining anchors are selected negative anchors, do you have a theory on why this do not cause any unbalanced training?

**Comment by Joker316701882** [comment]
@barisbatuhan 
> Do you automatically assign all remaining/unmatched anchors as negative samples?

Yes.

> number of positive and negative anchors are balanced with that way (like getting "num_positive_anchors x 3" times negative anchors). Which strategy is followed to select negative anchors and in which part of the YOLOX_Head class can I find that structure?

There are many ways to handle the imbalance issue between positive/negative anchors. For SSD/Faster R-CNN, they use a balanced sampling strategy as you mentioned. For RetinaNet/FCOS, they use focal loss. For the advanced YOLO series(YOLOv4/v5/x), we only initialize the classification bias so that at the beginning stage of the training process, most anchors are predicted as negative, this is originally proposed in the RetinaNet paper. Some researchers found this is enough for dealing with the imbalance issue.

**Comment by barisbatuhan** [comment]
Thank you very much, this was quite helpful for understanding the code

**Comment by feiyuhuahuo** [comment]
@Joker316701882   Have you tried other weight initialization method? And what's the difference?

> Some researchers found this is enough for dealing with the imbalance issue.

Is there any paper? I'm quite interested in it, thanks.

**Comment by barisbatuhan** [comment]
@feiyuhuahuo, As @Joker316701882 said, it is probably explained in the **RetinaNet** paper. You should check that first. About the weight initialization, if the weights are not initialized properly and posiitve/negative anchor ratio is not limited to a value, these happen:

1) There are overwhelmingly many negative anchors compared to positive anchors. Therefore, instead of learning positive anchors correctly, the model first tries to predict every anchor as negative, since this also reduces the loss to some extent. 
2) After the stage where "most anchors are predicted as negative", the model starts to learn the positive anchors (if it can, sometimes it just cannot since the model stucks in a local minima), since the loss cannot be reduced further by predicting all boxes more negative.

To overcome at least the first step (and to make the model trainable in the second step), the biases should be initialized properly. Instead of initializing like in this repository, I tried to initialize with the default settings of PyTorch, and the model did not function well.
